### PR TITLE
Don't sort batches during plan

### LINKED ISSUE #1939: SortExec No Longer Streams Correctly
**Describe the bug**

https://github.com/apache/arrow-datafusion/pull/1596/files#diff-68811b72d27f9f5173223e0da1af2a467c2e4fff2f5f2237665fa29e1a6575c0L165 appears to have accidentally changed the behaviour of `SortExec` so that it no longer returns a stream that performs the sort operation, but instead performs the sort within `ExecutionPlan::execute`.

This effectively stalls out constructing the rest of the physical plan until the sort has completed, and prevents result streaming from working correctly.

**To Reproduce**

Run a query with a large `SortExec`, observe surprising amount of time spent in `ExecutionPlan::execute`

**Expected behavior**

`ExecutionPlan::execute` should return a stream of results, but should not block on those results being available

**Additional context**

This resulted in what looked like missing traces in IOx (https://github.com/influxdata/influxdb_iox/issues/3822) as it never actually finished constructing the physical plan from which to collect metrics :sweat_smile: 

FYI @yjshen

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #1939.

 # Rationale for this change

Currently SortExec performs the sort during the query plan

# What changes are included in this PR?

This makes the SortExec lazy

# Are there any user-facing changes?

SortExec no longer computes values during query planning

### CODE REVIEW

File: datafusion/core/src/physical_plan/stream.rs [hunk 1]
```diff
@@ -73,3 +74,45 @@ impl RecordBatchStream for RecordBatchReceiverStream {
         self.schema.clone()
     }
 }
+
+/// Combines a [`BoxStream`] with [`SchemaRef`] implementing
+/// [`RecordBatchStream`] for the combination
+pub(crate) struct RecordBatchBoxStream {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/stream.rs:80]: I'm not a massive fan of this, as it results in double dyn-dispatch, but it is probably acceptable until such a time as we rework ExecutionPlan.

I've put this alongside RecordBatchReceiverStream as it is likely generally useful, but I've kept it pub(crate) as I'm not sure how long this interface will be around for

File: datafusion/core/src/physical_plan/sorts/sort.rs [hunk 2]
```diff
@@ -779,17 +779,21 @@ impl ExecutionPlan for SortExec {
...
-            context,
-        )
-        .await;
-
-        debug!("End SortExec::execute for partition {}", partition);
-        result
+        Ok(Box::pin(RecordBatchBoxStream::new(
+            self.schema(),
+            futures::stream::once(
+                do_sort(
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/sorts/sort.rs:785]: A nicer fix would be to make ExternalSort stream, or less async, or some combination of the two, but in the short term this is a quick fix

**Review by alamb — APPROVED** [review]

File: datafusion/core/src/physical_plan/stream.rs [hunk 3]
```diff
@@ -73,3 +74,45 @@ impl RecordBatchStream for RecordBatchReceiverStream {
         self.schema.clone()
     }
 }
+
+/// Combines a [`BoxStream`] with [`SchemaRef`] implementing
+/// [`RecordBatchStream`] for the combination
```
[the remark this answers is not in the document]
Comment (tustvold) [reply datafusion/core/src/physical_plan/stream.rs]: I've updated it to no longer need a boxed stream, the trade-off is more pinning nonsense :sweat_smile:   PTAL

File: datafusion/core/src/physical_plan/stream.rs [hunk 4]
```diff
@@ -73,3 +74,52 @@ impl RecordBatchStream for RecordBatchReceiverStream {
         self.schema.clone()
     }
 }
+
+pin_project! {
+    /// Combines a [`Stream`] with a [`SchemaRef`] implementing
+    /// [`RecordBatchStream`] for the combination
+    pub(crate) struct RecordBatchStreamAdapter<S> {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/stream.rs:81]: 👍

**Review by alamb — APPROVED** [review]
Looks better now (though I will be honest I don't really understand how `pin_project` works and how the code generated by this will be different than a `Box`d stream 🤷

**Comment by tustvold** [comment]
It'll have one-less stack allocation per stream, and one less dyn-dispatch per poll. Whether that matters 🤷‍♂️

**Comment by alamb** [comment]
> It'll have one-less stack allocation per stream, and one less dyn-dispatch per poll. Whether that matters 🤷‍♂️

What I was getting at was that I don't know what sort of magic occurs as part of `pin_project` there may be more allocations I am not aware of. I am not worried, just opining

**Comment by tustvold** [comment]
It's really just a trick to ensure a given field is consistently either structurally or non-structurally pinned - https://doc.rust-lang.org/nightly/core/pin/index.html#projections-and-structural-pinning. 

Unlike say async_trait, there isn't that much wizardry occuring under the hood, the generated code is actually relatively straightforward

```
pub(crate) fn project<'__pin>(
    self: ::pin_project_lite::__private::Pin<&'__pin mut Self>,
) -> Projection<'__pin, S> {
    unsafe {
        let Self {
            schema, stream
        } = self.get_unchecked_mut();
        Projection {
            schema: schema,
            stream
            : ::pin_project_lite::__private::Pin::new_unchecked(stream
            ),
        }
    }
}
```

**Comment by alamb** [comment]
> It's really just a trick to ensure a given field is consistently either s

TIL

**Pull request merged**